### ISSUE TITLE
Synchronizing beta releases with ocamlfind

### ISSUE THREAD

**Issue by maroneze** [ocaml/ocaml#8781]
I've just seen the announcement of the 4.09 beta and would like to try it, but unfortunately ocamlfind is still not operational in this version. Since almost all of my packages rely on it, there's not much I can do for now, except periodically poll `opam update`. But what usually happens is that I end up forgetting about it until the next beta release is announced.

I imagine several people are in the same situation, given the importance of ocamlfind, so I wonder if it would be better to synchronize beta releases with ocamlfind, or at least announce it on the mailing list, when ocamlfind is updated in opam.

**Comment by dra27** [comment]
There is indeed a critical set of packages which aid testing a beta release (there’s lots of ongoing work to try to reduce this frenzy). I’m not certain this wants coordinating from the OCaml dev team, though - a Discuss post highlighting the immediate problems and updated with details of patches/pins might be more appropriate than effectively a two-stage beta announcement from us?

**Comment by bluddy** [comment]
It seems to me like we need a new stage for OCaml releases, where the release leaves the hands of the core dev team, but is not ready for consumption by the average user until the main elements of the ecosystem/platform are in place. Having another team in place to take over at that point (perhaps piggybacking off of the ocaml-community organization) could be helpful with this.

**Comment by maroneze** [comment]
Indeed, maybe it could be simply some sort of informal agreement that each "major" package piggybacks on the release e-mail to notify people: "Just to let you know, <package> is now working on this release". I don't tend to go to Discuss very often, I prefer the push style of e-mails.

Anyway, Discuss is probably a better place for this than a Github issue, so I'm closing it. Thanks for your inputs.

**Comment by ejgallego** [comment]
I am not sure if a beta release may be very useful if `findlib` and `dune` are known not to work.

It seems to me that not a lot of packages would be able to be tested indeed. IMHO the pre-release check should include some basic sanity checks to ensure the key infrastructure does work.

**Comment by ejgallego** [comment]
By the way `findlib` was fixed a few months ago [as we need it for the Coq's CI test which runs against OCaml trunk`], so all that is needed there is a new release, or alternatively pinning against findlib's repo.

**Comment by dra27** [comment]
@ejgallego - that seems to be more an argument in favour of the release sitting in beta for as long as is necessary to allow key packages to stabilize, which is already the release procedure (we don't "sign off" a release if we're aware that there's too much breakage... a lesson we learned from 4.06)

**Comment by pmetzger** [comment]
> There is indeed a critical set of packages which aid testing a beta release (there’s lots of ongoing work to try to reduce this frenzy). I’m not certain this wants coordinating from the OCaml dev team, though

I've been thinking about it, and I now think central coordination is important. Without some formal coordination, releases are going to remain a mixed success. It's not enough to just put the beta out there and hope that the people maintaining crucial packages will notice and act; something like a mailing list and some tooling may be necessary to make sure everyone gets alerted and the work proceeds smoothly.

**Comment by ejgallego** [comment]
@dra27 my concern with that model is that it may imply that people testing the beta spend too many (redundant) cycles: they all download the beta in parallel, discover the breakage, report it, etc... as @pmetzger a more central coordination seems important to reduce the barriers to beta testing.

For Coq we follow a slightly different model, key packages are already in our own CI and `master` is always known to work with them. This implies that some PRs may get delayed until such packages are fixed, but on the other hand when we tag a beta many base key packages are known to work so beta testers can focus on their own developments.

**Comment by dra27** [comment]
@pmetzger - this point was made (https://github.com/ocaml/ocaml/issues/8781#issuecomment-507354319 and https://github.com/ocaml/ocaml/issues/8781#issuecomment-507356744). The fact that `findlib` has been up-to-date but not everyone knew about the pin confirms it would be a good idea. *But not here!*

@ejgallego - agreed on the redundant cycles; I wonder if opam may be able to assist with this. For the other packages, we used to test certain packages with each PR (`camlp4` and `ocamlbuild`, I can't remember if there were ever others). However, we stopped doing that inasmuch as it's not necessarily (indeed, often) the job of the PR author to fix the packages. We also actively discourage development against `trunk` (I know Coq is an exception!). We are working on other tooling to detect these breakages sooner, but CI isn't the answer here.

**Comment by ejgallego** [comment]
Thanks for the information @dra27 ! I was not aware that testing against OCaml `trunk` was not recommended, is there something such as release policy document?

In fact I had the opposite idea, given for example what the plans for PPX are, which aim to help PPX work in trunk.

In Coq, we started testing against `trunk` as indeed we had a couple of instances where Coq would be broken in newer OCaml versions. IIRC these were OCaml bugs, however would we had not tested against the beta maybe the release would have been bugged for us.

Even so, testing the beta brought quite a bit of latency to the process, so we prefer to catch these kind of problems as early as possible.

[I'm taking the freedom to CC a few Coq devs who are interested in release processes @zimmi48 @skyskimmer @maximedenes]

**Comment by dra27** [comment]
Not recommended doesn't mean never do it - it's just that the stability guarantees of `trunk` are not as high as for a release branch. For example, we do occasionally cope (with great irritation) at some platforms being broken in `trunk` for a period of time (the irritation is that we get peppered with CI failure warnings on every merge, which acts as an incentive to fix the problem). So for Coq, where you're happy to react and prepare for changes in OCaml trunk quickly, that works fine but for the majority of package maintainers, it's probably not so appropriate.

**Comment by dra27** [comment]
I'm not sure that we have anything written down about the branches in the sense of what the appropriate usage is - @gasche?

**Comment by Zimmi48** [comment]
Just FTR our trunk test is in allow failure mode which means that we don't have to react *that* quickly to new breakage.

But overall, it would seem like a good idea for the OCaml dev team to inspire from the Coq CI process. @ejgallego and @SkySkimmer: We should write a paper about it so that what we consider as valuable lessons can be shared with other projects like OCaml.

**Comment by dra27** [comment]
@Zimmi48 - I will, with my tongue firmly in my cheek, point out that Coq's **master** branch CI has apparently been down for 3 days :wink:

More seriously, what seems to be missing in the "what we can learn" is what you expect to change in OCaml. We're not going to delay merging PRs in OCaml because it breaks downstream packages - we've already done that in the past, and the effect is to stall features, because it results in branches going stale and requiring (painful) rebases. It's also our experience that "allowed fail" CI tests are a waste of energy, because noone ever looks at them (for us, at least, that doesn't mean they are for everyone - and the fact they are not looked at is more a UI problem than an issue with the idea of an allowed-fail test).

At the PR against OCaml's trunk which broke findlib, what precise steps would you like to have happened, and who would have carried them out?

**Comment by ejgallego** [comment]
> has apparently been down for 3 days

CI on Coq master is fine, for example https://gitlab.com/coq/coq/pipelines/71186436 , the "windows" workers are indeed not properly maintained. Keep in mind that we have 0 "official" resources towards CI, so it is indeed not easy to maintain. Overall, the buildability of master has been pretty good.

> we've already done that in the past, and the effect is to stall features, because it results in branches going stale and requiring (painful) rebases.

That's interesting, it was not our experience, tho we require every package in the CI to take our patches. On the other hand we are trying to implement a very strict compatibility policy so YMMV.

> It's also our experience that "allowed fail" CI tests are a waste of energy,

We use them as "warnings", for example, compilation with OCaml trunk is broken. I'd dare to say it has worked well to keep us in the safe side for a while.

> At the PR against OCaml's trunk which broke findlib, what precise steps would you like to have happened, and who would have carried them out?

So in this case, findlib would have been discovered to fail, then a PR against findlib would have been submitted, and the PR would have been integrated. This is actually what happened when we got the warning in Coq. This does indeed not fix the problem in OPAM, but we are working on the Coq platform that would add the "release a compatible package" step into the workflow.

**Comment by gasche** [comment]
What I personally want to have in OCaml is a CI that constantly builds the opam world from trunk (preferably something that I can run myself on a machine somewhere), and lets me know when things get broken. When a given package break, it gives myself one week to investigate the failure, decide if it was a regression in trunk and fix it, or decide that the package needs to be updated. The second week, it notifies the package managers of the issue, which hopefully they or I can fix during the week, in which case I pin their trunk-support branch to be able to keep building.

**Comment by dbuenzli** [comment]
@gasche the problem is that you will have difficulties in discriminating between failure that are due to trunk and those from the (non-)health of the opam repository. 

With respect to the specific problem mentioned by the OP, I think it just shows it may not have been such a good idea to provide so little infrastructure support and conventions in the toolchain that using projects with third-party libraries requires a third-party tool which brings in its own indirection and new (uneeded) concepts and names.

Maybe that's the real problem that needs fixing... and by that I don't mean integrating `ocamlfind` in `ocaml`, it has already been shown you don't really need it since the compilation objects largely self describe their dependencies.

**Comment by nojb** [comment]
Off-topic, but my feeling is that ocamlfind is on its way out with the advent of dune.

**Comment by dbuenzli** [comment]
I neither think `dune` solved any problem on that front nor that it is the right place to fix them.

**Comment by dra27** [comment]
Let’s stay on topic for ocaml/ocaml, please.

@gasche - @kit-ty-kate’s opam health check tool is probably the fit here and because it shows the changes from the previous build, it would go some way to ameliorate @dbuenzli’s valid point about repo health noise. But to be useful, we need the ppx advances presently being worked on so that more packages work with trunk more often

**Comment by ejgallego** [comment]
> What I personally want to have in OCaml is a CI that constantly builds the opam world from trunk ...

The main problem with this approach @gasche , is that there two weeks all the other developers will see the failure due to the non-consistency, making the whole system not useful to them, right? Or are you thinking of a different approach [I understood you were talking about `trunk`, that is, already merged PRs]

> But to be useful, we need the ppx advances presently being worked on so that more packages work with trunk more often

Note that IMHO it is possible to update [some core] ppx to keep working in trunk as we do for example in Coq plugins, the key to this is IMHO having the right tooling so you can do a composed build of all the overlays and of course the overlay creation / pushing scripts.

[In Coq an overlay is a layer (or "pin") or a third party package that accompanies a PR, when the PR is merged the overlay is merged in the corresponding `master` branch of the development]

**Comment by kit-ty-kate** [comment]
PSA: a release of ocamlfind has been made an hour ago and will be available shortly in opam: https://github.com/ocaml/opam-repository/pull/14603
